Simplify Python examples and use JupyterLab (#129)
HealthBioscienceIDEAS/microscopy-novice: episodes/instance-segmentation-and-measurements.md

Reviewer: ```suggestion
'Intensity Image Layer' drop down box. You can choose to measure various shape
```
Just noticed a missing space here between 'drop down box.' and 'You'

Reviewer: As this adds more content, could you update the teaching time estimate here? It's in minutes

I'd probably go for at least 1hr teaching time, then maybe 20 minutes for exercises? That total of 1hr20 matches the size of the session after the lunch break which sounds about right.

Author: Yep, I think that is about right. I've updated the time estimates.

- Use connected components labelling on a thresholded image.
- Calculate the number of cells and average cell volume.
- Save and edit your workflow to reuse on subsequent images.
- Perform more complex cell shape analysis using scikit-image's `regionprops`.

::::::::::::::::::::::::::::::::::::::::::::::::

In this lesson we'll continue to work with the Cells (3D + 2Ch) image we've
been using in past lessons. 

Instead of entering Python commands in Napari’s built‑in console, we will write and run our code in a [computational notebook](https://docs.jupyter.org/en/latest/#what-is-a-notebook) using [JupyterLab](https://jupyterlab.readthedocs.io/en/stable/index.html). This allows us to build a reusable workflow that is easy to repeat and adapt.

We have kept the level of programming knowledge required to the minimum
possible and all code can be run by copy and pasting, so don't worry if
you don't understand it all yet.

Most, if not all, of the functions we will use in this lesson are also
accessible via various Napari plugins, so the analysis pipeline could also be
assembled within Napari if you prefer.

## Creating a notebook in JupyterLab

### 1. Activate your Napari environment

> Open the Miniforge Prompt (or whichever other environment manager you used when installing Napari), and activate the environment you created for Napari:

Reviewer: ```suggestion
Open the terminal (the same one you used for Napari installation: see 'Opening a terminal' section of [the setup instructions](http://127.0.0.1:4321/index.html#install-python-packages)), and activate the environment you created for Napari:
```
Even when participants are using the same environment manager (miniforge), they will access it in different ways on different operating systems (miniforge prompt for windows, standard terminal for mac / linux etc). Maybe make this more generic by referring to a terminal + linking to the setup instructions?

Author: Yes, good point!

This will open a new Python notebook.

### 5. Name your notebook
Renaming your notebook immediately helps keep your workflow tidy and makes it easier to find later.

Right click the default name at the top of the notebook tab (e.g., *Untitled.ipynb*).

Enter a meaningful name, for example: *instance_segmentation.ipynb*.

## About notebooks
A notebook is made up of building blocks called **cells**.

For this workshop, we will only use **Code Cells**. 

When you run a Code Cell, the output typically appears underneath it. This could be a number, text, a table, an error message.

By splitting code up into cells, you can run one specific part of your code without having to re-run the whole file and get instant feedback.

Be careful about the order of your notebook cells. Running them out of sequence can leave variables outdated or missing, which can lead to confusing results. 

## Using Python inside a notebook

Run each of the following examples in separate notebook cells so you can clearly see the output after each step.

``` python
# Everything after a hash (#) is a comment and is ignored by Python.
# Use comments to explain what you're doing.
```

> If you want to create another cell, click the **+** button in the toolbar or use **Insert > Insert Cell Below**.

Reviewer: I couldn't see an `Insert > Insert Cell Below` option in the jupyter lab menus. Is this referring to the button within the cell on the right side (the box with a plus below it)? If so, might need to be a bit more explicit about this to help people find it - something like:

```suggestion
If you want to create another cell, click the **+** button in the toolbar or use the **Insert Cell Below** button on the right side of the cell.
```

Author: Hmm, yeah I think that must have been what I meant!

# Python can assign values to variables
one = 1
two = one + one
```

Notice that there is no output.

``` python
# Variables store values rather than return them
# To see their value write the variable name
two
```

``` output
2
```

**Note**: In a standard Python script, writing a variable name on its own does nothing and you must use `print()` to show output.

``` python
# Python's print function
print("one plus one is", two)
print("one plus two is", one + two)
```

``` output
one plus one is 2
one plus two is 3
```

## Using Napari from within a notebook

``` python
# Import napari package
import napari
``` 

``` python
# Open Cells (3D + 2Ch) sample image in napari's viewer
viewer = napari.Viewer()
viewer.open_sample("napari", "cells3d")
```

The output should look like this:

``` output
[<Image layer 'membrane' at 0x1853b7738c0>,
 <Image layer 'nuclei' at 0x1853c844710>]
 ```

The memory addresses (`0x1853b7738c0` and `0x1853c844710`) will be different for you. They indicate the locations in memory where Python happened to store those layer objects. 

Napari's viewer should open in a separate window, preloaded with the cells3D sample image.

## Semantic segmentation

> 

Reviewer: Might be worth adding a bit of additional context here, that we're reproducing what we did in the previous episode e.g. something like:

First, we will re-create the semantic segmentation of our nuclei that we made in the previous episode.

- How do we perform instance segmentation in Napari?
- How do we measure cell size with Napari?

> - How can computational notebooks be used to build re‑usable workflows?

Reviewer 2: Elsewhere re-use has become reuse, so for consistency ... 
```suggestion
- How can computational notebooks be used to build reusable workflows?
```
What does codespell say?

- Save and edit your workflow to reuse on subsequent images.

- Perform more complex cell shape analysis using scikit-image's `regionprops`.

::::::::::::::::::::::::::::::::::::::::::::::::

In this lesson we'll continue to work with the Cells (3D + 2Ch) image we've
been using in past lessons. 

Instead of entering Python commands in Napari’s built‑in console, we will write and run our code in a [computational notebook](https://docs.jupyter.org/en/latest/#what-is-a-notebook) using [JupyterLab](https://jupyterlab.readthedocs.io/en/stable/index.html). This allows us to build a reusable workflow that is easy to repeat and adapt.

We have kept the level of programming knowledge required to the minimum
possible and all code can be run by copy and pasting, so don't worry if
you don't understand it all yet.

Most, if not all, of the functions we will use in this lesson are also
accessible via various Napari plugins, so the analysis pipeline could also be
assembled within Napari if you prefer.

## Creating a notebook in JupyterLab

### 1. Activate your Napari environment
Open Miniforge Prompt and activate the environment you created for Napari:
``` bash

> activate napari-env

Reviewer: ```suggestion
conda activate napari-env
```

Author: good catch

```
### 2. Create and navigate to your workshop folder

> It is best practice to keep all your project files together in a dedicated folder.

Reviewer: I wonder if we want something here about moving to a location they can find again (like /Documents, /Desktop) to create this folder? At least on windows miniforge prompt opens by default in: `C:\Users\username` which is tough to find again through the standard file browser.

Potentially, we could do this moving / folder creation via the jupyter interface to avoid multiple bash commands. E.g. if they open `juptyer lab` in the default location, they could then use the file browser at the left side of jupyter to move to a sensible location and create a new folder.  What do you think?

Author: Yep this makes sense. I've adjusted the instructions.

``` bash
# Create a folder
mkdir napari-workshop

# Move into that folder
cd napari-workshop
```
Everything you create in JupyterLab will be saved inside this folder.

### 3. Launch JupyterLab
Start JupyterLab from inside the napari-workshop folder:
``` bash
jupyter lab
```
The JupyterLab interface should appear in a browser window.

### 4. Create a new notebook

> In the JupyterLab Launcher, **select: Notebook > Python 3 (ipykernel)**

Reviewer 2: For me this was 

```suggestion
In the JupyterLab Launcher, **select: File > New > Notebook > Python 3 (ipykernel)**
```

This will open a new Python notebook.

### 5. Name your notebook
Renaming your notebook immediately helps keep your workflow tidy and makes it easier to find later.

> Right click the default name at the top of the notebook tab (e.g., *Untitled.ipynb*).

Reviewer: ```suggestion
Right click the default name at the top of the notebook tab (e.g., *Untitled.ipynb*), and select `Rename Notebook...`.
```

Enter a meaningful name, for example: *instance_segmentation.ipynb*.

## About notebooks
A notebook is made up of building blocks called **cells**.

For this workshop, we will only use **Code Cells**. 

> When you run a Code Cell, the output typically appears underneath it. This could be a number, text, a table, an error message.

Reviewer: ```suggestion
When you run a Code Cell, the output typically appears underneath it. This could be a number, text, a table, or an error message.
```

> By splitting code up into cells, you can run one specific part of your code without having to re-run whole file and get instant feedback.

Reviewer 2: ```suggestion
By splitting code up into cells, you can run one specific part of your code without having to re-run the whole file and get instant feedback.
```

Be careful about the order of your notebook cells. Running them out of sequence can leave variables outdated or missing, which can lead to confusing results. 

## Running code and adding new cells

> Write this in the first cell and run the cell:

Reviewer 2: This may not be obvious to complete beginners. For the workshop you can demonstrate but we could include a little graphic here also. 
```suggestion
Write this in the first cell and run the cell by pressing the right arrow in the menubar:
```

``` python
# Everything after the hash is a comment and won't be treated as code

# Python can do basic arithmetic
1 + 1
```

The result will appear directly underneath the cell.

If you want to create another cell, click the **+** button in the toolbar or use **Insert > Insert Cell Below**.

## Using Napari from within JupyterLab 
In the first cell run:
``` python
# Import napari package
import napari
``` 
In the second cell run:
``` python
# Open Cells (3D + 2Ch) sample image in napari's viewer
viewer = napari.Viewer()
viewer.open_sample("napari", "cells3d")
```

> A Napari viewer window opens in a separate window.

Reviewer 2: ```suggestion
A Napari viewer window should open in a separate window, preloaded with the cells3D sample image we have used in previous episodes```

## Create a mask
```python
# Create a semantic segmentation using scikit-image

# Import the functions we need from scikit-image
from skimage.filters import threshold_otsu, gaussian

# Access the nuclei channel from the viewer
image = viewer.layers["nuclei"].data

# Smooth the image and compute a threshold
blurred = gaussian(image, sigma=3)
threshold = threshold_otsu(blurred)

# Create a binary mask
semantic_seg = blurred > threshold

# Add the mask as a labels layer in Napari
viewer.add_labels(semantic_seg)
```

![](fig/semantic-seg-napari.png){alt="A screenshot of a rough semantic
segmentation of nuclei in Napari"}

In the Napari viewer you should see the image above. 

> You are now ready to begin this lesson.

Reviewer 2: I feel like the lesson has already begun. Maybe ..
```suggestion
We will now use Jupyter to make measurements with repeatable, well documented scripts. 
```

objects. Then it will assign each connected region a unique integer value.
Let's try it.

Create a new cell and run:

```python

> # Create an instant segmentation

Reviewer 2: ```suggestion
# Create an instance segmentation
```

![](fig/region_props_after.png){alt="A screenshot of the numeric value table
created by the napari-skimage plugin"}

## Regionprops

> Before, we used the napari‑skimage plugin to create a table with properties of the nuclei. The same properties can also be computed using the skimage directly in our notebook.

Reviewer: ```suggestion
Before, we used the napari‑skimage plugin to create a table with properties of the nuclei. The same properties can also be computed using scikit-image directly in our notebook.
```

# Import tools
from skimage.measure import regionprops_table
import pandas as pd

# Compute region properties
props = regionprops_table(
    label_image=instance_seg,
    properties=["label", "area", "centroid"]
)

# Convert to a pandas DataFrame
props_df = pd.DataFrame(props)

# Display the table
props_df
```

## Sorting and inspecting the results

Regionprops can generate a lot of information on the shape and
size of each connected region. For now we will focus only on the
column headed `area`, which shows the size in pixels.

Let's sort our table so that it is easier to see the extreme values.

``` python
# Sort the table based on cell size (area)
sorted_props_df = props_df.sort_values("area")

# Display the table
sorted_props_df
```

### The largest nucleus
According to the table, nucleus 3 is larger than the other nuclei
(202258 pixels). In the [what is an image](what-is-an-image.md#pixels)
lesson, we learnt to use the mouse pointer to find particular values in an
image. Hovering the mouse pointer over the light purple nuclei at the bottom
left of the image we see that these apparently four separate nuclei have
been labelled as a single nucleus. 

> In the layer controls of the semantic_seg layer we can confirm this by selecting label `3` and enabling `show selected`.

Reviewer: ```suggestion
In the layer controls of the instance_seg layer we can confirm this by selecting label `3` and enabling `show selected`.
```

:::::::::::::::::::::::::challenge
### Why Are Separate Nuclei Getting the Same Label?

![](fig/same_label_2d.png){
alt="A screenshot of an instance segmentation of nuclei."}

> the images above, three of the light purple nuclei are visibly touching, so

Reviewer: ```suggestion
In the image above, three of the light purple nuclei are visibly touching, so
```

){alt="Napari's 2D/3D toggle" height='30px'} button.
Try it now.

![](fig/same_label_3d.png){
alt="A screenshot of an instance segmentation of nuclei in 3D mode with some
incorrectly joined instances."}
You should see the image rendered in 3D, with a clear join between the
upper most light purple nucleus and its neighbour.

::::::::::::::::::::::::: 
::::::::::::::::::::::::: 

### The smallest nucleus

> The smallest nucleus is labelled 18, with a size of 7 pixels. We can use the position data (the `centroid` columns) in the table to help find this nucleus. We need to navigate to slice 33 and get the mouse near the top left corner (33 64 0) to find label 18 in the image.

Reviewer: ```suggestion
The smallest nucleus is labelled 18, with a size of 7 pixels. We can use the position data (the `centroid` columns) in the table to help find this nucleus. We need to navigate to slice 33 and get the mouse near the top left corner (33 63 0) to find label 18 in the image.
```
64 also works, but as the table lists the centroid as (33.0000, 62.714 0.0000) might make sense to use the closest rounded value (62.714 -> 63)

![](fig/region_props_after_18.png){alt="A screenshot
region-props dialog highlighting the smallest nucleus."}

Nucleus 18 is right at the edge of the image, so is only a partial nucleus. Partial nuclei will need to be excluded from our analysis. We'll do this later in the lesson with a [clear border](#removing-border-cells) filter. However, first we need to solve the problem of joined nuclei.

## Separating joined nuclei
Our first problem is how to deal with four apparently distinct nuclei (labelled
with a light purple colour) being segmented as a single nucleus. 

### Erosion

> In order to use the label function to count the cell nuclei we first need

Reviewer: I think the explanation here could be simplified a bit. Here's a suggestion, but feel free to edit:

To separate our nuclei, we can 'erode' our segmentation. Erosion is a type of filter, similar to those we covered in the [filters and thresholding episode](filters-and-thresholding.md). It will make all segmented nuclei smaller, by setting pixels at their edge to zero.

The size / shape of the region that gets set to zero is controlled by the filter's 'footprint'. We'll use scikit-image's `ball` function to generate a sphere to use as the footprint. Any pixels closer to the edge of the nucleus than the radius of this sphere will be set to zero.

Create a new cell and run:

Try using a Python `for` loop to test several radius values.

:::::::::::::::::::::::::solution
You can change the radius manually (for example, `radius = 5`) and re‑run the erosion each time.  
But if you want to test **many** radius values quickly, a Python [for loop](https://swcarpentry.github.io/python-novice-inflammation/05-loop.html) lets you repeat the same steps for each radius in a list.

```python
# List of radii to test
radii = [5, 10, 15]
# A for-loop that tests several radii
for radius in radii:
    # Make a name for the output layer
    layer_name = "eroded_ball_" + str(radius)
    # Erode the mask using this radius
    eroded_mask = erosion(semantic_seg, footprint=ball(radius))
    # Add the eroded mask as a new layer in Napari
    viewer.add_labels(eroded_mask, name=layer_name)
```
:::::::::::::::::::::::::
:::::::::::::::::::::::::

It is possible to run the erosion function through a plugin: `Layers > Filter > Morphology > Binary Morphology (napari skimage)`.

### Instance segmentation using the eroded mask

Now we have separate nuclei, lets try creating instance labels
again.

```python
# Create a new instance segmentation using the eroded mask

> eroded_mask = viewer.layers['eroded_ball_10'].data

Reviewer: If people skipped the exercise above, or tried different values for the radius - they may not have 'eroded_ball_10' as a layer in the napari viewer. Probably best to have the erosion code directly here so all can follow along?
```suggestion
eroded_mask = erosion(semantic_seg, footprint=ball(10))
```

instance_seg = label(eroded_mask)

# Remove old instance segmentation
viewer.layers.remove('instance_seg')

# Add new instance segmentation
viewer.add_labels(instance_seg)
```

![](fig/instance_segmentation_on_eroded_mask.png){
alt="Instance segmentation on the eroded segmentation mask"}

Looking at the image above, there are no longer any incorrectly joined nuclei. 

### Dilation

We managed to separate the nuclei, however performing any size or shape analysis on these nuclei will be flawed, as they are heavily eroded.

> We can largely undo the erosion by using the scikit-image's [expand labels](

Reviewer: ```suggestion
We can largely undo the erosion by using scikit-image's [expand labels](
```

Try using a Python `for` loop to test several radius values.

:::::::::::::::::::::::::solution
You can change the radius manually (for example, `radius = 5`) and re‑run the erosion each time.  
But if you want to test **many** radius values quickly, a Python [for loop](https://swcarpentry.github.io/python-novice-inflammation/05-loop.html) lets you repeat the same steps for each radius in a list.

```python
# List of radii to test
radii = [5, 10, 15]
# A for-loop that tests several radii
for radius in radii:
    # Make a name for the output layer
    layer_name = "eroded_ball_" + str(radius)
    # Erode the mask using this radius
    eroded_mask = erosion(semantic_seg, footprint=ball(radius))
    # Add the eroded mask as a new layer in Napari
    viewer.add_labels(eroded_mask, name=layer_name)
```
:::::::::::::::::::::::::
:::::::::::::::::::::::::

### Instance segmentation using the eroded mask

Now we have separate nuclei, lets try creating instance labels
again.

```python
# Create a new instance segmentation using the eroded mask
eroded_semantic_seg = viewer.layers['eroded_ball_10'].data
eroded_instance_seg = label(eroded_semantic_seg)
viewer.add_labels(eroded_instance_seg)

> print("Number of nuclei after erosion  =", instance_seg.max())

Reviewer 2: I think this should be ...
```suggestion
print("Number of nuclei after erosion  =", eroded_instance_seg.max())
```

:::::::::::::::::::::::::
:::::::::::::::::::::::::

## Removing Border Cells
Now we return to the second problem with our initial instance segmentation,
the presence of partial nuclei around the image borders. As we're measuring
nuclei size, the presence of any partially visible nuclei could substantially
bias our statistics. 

We can remove these from our analysis using scikit-image's [clear border](
https://scikit-image.org/docs/stable/api/skimage.segmentation.html#skimage.segmentation.clear_border) function.

```python
# Remove partial nuclei touching the image border

# Import scikit-image's clear_border
from skimage.segmentation import clear_border

# Clear border
instance_seg = clear_border(instance_seg)

# Remove old instance segmentation
viewer.layers.remove('instance_seg')

# Add new instance segmentation
viewer.add_labels(instance_seg)
```

![](fig/instance_segmentation_clear_border.png){
alt="The instance segmentation with any nuclei crossing the image boundary
removed"}

> We now have an image with 11 clearly labelled nuclei. You may notice that the smaller nucleus (dark orange) near the top left of the image has been removed even though we can't see where it touches the image border. Remember that this is a 3D image and clear border removes nuclei touching any border. This nucleus has been removed because it touches the top or bottom (z axis) of the image. Let's check the nuclei count as we did above.

Reviewer: You end this paragraph with 'Let's check the nuclei count as we did above' - but all examples of using `.max()` to retrieve the nuclei count have been removed prior to this point.

So either you need to include an earlier example of using `.max()` to get the nuclei count, or explain it from scratch here. E.g. there used to be a `.max()` call in the [counting the nuclei section](https://healthbioscienceideas.github.io/microscopy-novice/instance-segmentation-and-measurements.html#counting-the-nuclei)

Author: I added back the counting of the nuclei after creating the first instance segmentation

```python
# First count the nuclei

> number_of_nuclei = clear_border_dilated_instance_seg.max()

Reviewer: ```suggestion
number_of_nuclei = instance_seg.max()
```
Above the clear border image is saved into `instance_seg`, so it will need to match here too

print("Number of nuclei: ", number_of_nuclei)
```
```output
Number of nuclei: 19
```

Why are there still 19 nuclei? 

When we ran `clear_borders` the pixels corresponding
to border nuclei were set to zero, however the total number of labels
in the image was not changed, so whilst there are 19 labels in the
image some of them have no corresponding pixels. The easiest way to
correct this is to relabel the image (and replace the old instance
segmentation in the viewer.)

```python
# Relabel
instance_seg = label(instance_seg)

# Remove old instance segmentation
viewer.layers.remove('instance_seg')

# Add relabeled instance segmentation
viewer.add_labels(instance_seg)

# Number of nuclei after relabling
number_of_nuclei = instance_seg.max()
print("Number of nuclei:", number_of_nuclei)

```

```output
Number of nuclei: 11
```

> ## Number of pixels per nuclei.

Reviewer: ```suggestion
## Number of pixels per nucleus
```

Now that your instance segmentation is correct, you can finish the analysis in our notebook.

Let's start by counting the pixels per nucleus like we did before. 

```python
# Count the pixels per nucleus

# Extract region properties 
props = regionprops_table(
    instance_seg,
    properties=["label", "area"]   # 'area' = number of pixels
)

# Convert to a pandas DataFrame
props_df = pd.DataFrame(props)

props_df
```

> **Are these pixel counts good measurements?**

Reviewer: ```suggestion
**Are these pixel counts useful measurements?**
```
Maybe useful instead? For me, good implies whether the measure is accurate or not - we are measuring the pixel counts correctly, it's just that it doesn't mean much biologically

> Pixel counts depend on image resolution, not on the biology of the sample.

Reviewer: I'd maybe expand on this slightly:
```suggestion
Pixel counts depend on image resolution, rather than the real size of a biological structure. This means images of the exact same nuclei taken with different settings could give vastly different values for the number of pixels. 
```

Author: Yes, much better!

This is why biologists convert pixel counts into physical units like µm³ that allow comparisons across experiments, microscopes, and labs.

## Volume per nuclei

To convert to volumes we need to know the pixel size. 

In the lesson on [filetypes and metadata](filetypes-and-metadata.md#pixel-size) we learnt how to inspect the image metadata to determine the pixel size. 

Unfortunately the sample image we're using in this lesson has no metadata. Fortunately the image pixel sizes can be found in the [scikit-image documentation](https://scikit-image.org/docs/stable/api/skimage.data.html#skimage.data.cells3d). So we can assign a pixel size of 0.26&mu;m (x axis), 0.26&mu;m (y axis) and 0.29&mu;m (z axis).

Using this pixel size, we can then calculate the nucleus volume in cubic micrometres.

```python
# Volume of a single voxel in cubic micrometres
voxel_volume = 0.26 * 0.26 * 0.29

# Add a physical volume column
props_df["volume_um3"] = props_df["area"] * voxel_volume

props_df
```
Once you know the voxel size, `pandas` makes the conversion and analysis extremely easy:
```python
# Quick stats using pandas

> props_df["volume_um3"].describe()

Reviewer: This is a great shortcut for getting the summary stats! I'd forgotten about `.describe`

```
 
## Saving, reusing, and sharing your workflow

A key advantage of using a JupyterLab notebook is that your entire analysis is saved in one place. This makes your workflow reproducible, easy to adapt, and simple to share with others.

A tidy notebook is easier to understand for others (and for your future self). 

> Good practice include:

Reviewer: ```suggestion
Good practice includes:
```

- Ensuring the notebook runs without issues from beginning to end.

- Organising the notebook into clear sections (e.g. imports, loading data, segmentation, measurements, exporting results).

> - Removing unused cells and tidy temporary experiments.

Reviewer: ```suggestion
- Removing unused cells and tidying temporary experiments.
```

- Adding short notes explaining key steps and decisions.

> Your notebook contains all the analysis steps, but it won’t run correctly unless the software environment is the same. 

Reviewer: This is a good point, but I wonder if showing the `conda env` export commands is too much for this beginners lesson? Some points:
- If we want to include this, we'd need to be clear about where to run it i.e. in the terminal once the notebook has been closed. (I imagine people will try running it in jupyter notebook)
- These files often don't work for installation across different operating systems - as they may include platform-specific dependencies (e.g. if someone created it on windows, someone else probably couldn't use it to make the same env on a mac)
- There is [`--from-history`](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#exporting-an-environment-file-across-platforms) which can help with this; but it doesn't seem to work with our `napari-env` 

Maybe remove the specific commands and just link to the documentation? What do you think?

Author: Hmm, yep I agree. During the pilot of this material I also ended up skipping these details. I think you are right that it is too much for beginners (and also too much information wise, and too different from the previous content).

Author: I have changed it to:
> Your notebook contains all the analysis steps, but it won’t run correctly unless the software environment is the same. You can share the instructions you used to create your conda environment, but this doesn’t always reproduce the exact same package versions or dependencies. Conda also provides tools for exporting an environment so it can be recreated more reliably elsewhere. See the [conda documentation](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-from-an-environment-yml-file) for more information.

A tidy notebook is easier to understand for others (and for your future self). 

Good practice includes:
- Ensuring the notebook runs without issues from beginning to end.
- Organising the notebook into clear sections (e.g. imports, loading data, segmentation, measurements, exporting results).
- Removing unused cells and tidy temporary experiments.
- Adding short notes explaining key steps and decisions.

Your notebook contains all the analysis steps, but it won’t run correctly unless the software environment is the same. 

Export your conda environment so others can recreate it:

``` bash
conda env export > environment.yml
```
To recreate the same software environment on another computer:

``` bash
conda env create -f environment.yml
```
Read the [conda documentation](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-from-an-environment-yml-file) for more information.

Other good practices include:
- Sharing your notebook on a collaborative platform such as [GitHub](https://github.com/) or [GitLab](https://about.gitlab.com/), where others can comment, discuss, or propose improvements.

> - Using version control (e.g., [Git](https://git-scm.com/)) to track changes, document improvements, and maintain a clear history of how your workflow evolves. This also gives you the ability to revert to earlier versions whenever a mistake happens, ensuring you never lose working progress.

Reviewer 2: Typo?
```suggestion
- Using version control (e.g., [Git](https://git-scm.com/)) to track changes, document improvements, and maintain a clear history of how your workflow evolves. This also gives you the ability to revert to earlier versions whenever a mistake happens, ensuring you never lose work in progress.
```
